webrtc/README.md

```diff
@@ -0,0 +1,268 @@
+**Table of Contents**
+
+- [WebRTC](#webrtc)
+    - [Motivation](#motivation)
+    - [Requirements](#requirements)
+    - [Addressing](#addressing)
+    - [Connection Establishment](#connection-establishment)
+        - [Browser to public Server](#browser-to-public-server)
+            - [Open Questions](#open-questions)
+        - [Browser to Browser](#browser-to-browser)
+            - [Open Questions](#open-questions-1)
+    - [Connection Security](#connection-security)
+        - [Open Questions](#open-questions-2)
+    - [Multiplexing](#multiplexing)
+    - [General Open Questions](#general-open-questions)
+    - [Previous, ongoing and related work](#previous-ongoing-and-related-work)
+- [FAQ](#faq)
+
+<!-- markdown-toc end -->
+
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Addressing
+
+WebRTC multiaddresses are composed of an IP and UDP address component, followed
+by `/webrtc` and a multihash of the certificate that the node uses.
+
+Examples:
+- `/ip4/1.2.3.4/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+- `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+As a preparation browser _A_ [generates a
+certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate)
+and [gets the certificate's
+fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ instantiates a `RTCPeerConnection`, passing its local certificate as a
+   parameter. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_. _B_ learns of _A_ TLS fingerprint
+   through A's STUN message USERNAME attribute. See Open Questions below for
+   potential better solutions.
```

Comment:
It's possible to disable fingerprint verification on the server side. But then DTLS connection won't be secure, no?

Reply:
> It's possible to disable fingerprint verification on the server side.

Does webrtc-rs expose the fingerprint of the remote, that it learned through the DTLS handshake, to the user (here rust-libp2p) once the handshake is done @melekes?

Reply:
> Does webrtc-rs expose the fingerprint of the remote, that it learned through the DTLS handshake, to the user (here rust-libp2p) once the handshake is done @melekes?

https://github.com/webrtc-rs/webrtc/blob/4a5594e873319ef45d9ffb9fd9049dee18288503/src/dtls_transport/mod.rs#L451

Based on what I see, yes 👍 . _NOTE: Currently, it sets `remote_certificate` only when `disable_certificate_fingerprint_verification` is off (i.e. we validate fingerprint), but I think I can change it to set it under all conditions._

> If so, we would no longer need to embed the fingerprint in the STUN message. Instead, we can use the remote's fingerprint learned through the DTLS handshake in the following Noise handshake.

Okay, but what should I use for STUN username and password? 

- If _A_ constructs SDP it pretends to have received from _B_, then it needs _B_'s ICE username and password. Currently they are both set to _B_'s fingerprint. _(We need them to establish ICE connection)_
- _B_ can use username from STUN binding request (it can be random, BUT we don't know the password, so we'll have to agree that username==password).

So, while removing the necessity to put fingerprint into STUN binding request simplifies stuff (not a lot, I'd say it complicates it actually because one side will be putting fingerprints, another - random values), it does not really solve the problem described in https://github.com/libp2p/specs/pull/412#issuecomment-1170906786.

Reply:
> > Does webrtc-rs expose the fingerprint of the remote, that it learned through the DTLS handshake, to the user (here rust-libp2p) once the handshake is done @melekes?
> 
> https://github.com/webrtc-rs/webrtc/blob/4a5594e873319ef45d9ffb9fd9049dee18288503/src/dtls_transport/mod.rs#L451
> 
> Based on what I see, yes +1 . _NOTE: Currently, it sets `remote_certificate` only when `disable_certificate_fingerprint_verification` is off (i.e. we validate fingerprint), but I think I can change it to set it under all conditions._

That is great.

@ckousik can you confirm that the same is the case for Go's Pion, i.e. that it (a) allows an _insecure_ DTLS handshake and (b) exposes the fingerprint of the remote learned in the DTLS handshake to the user (here go-libp2p)?

> __
> 
> > If so, we would no longer need to embed the fingerprint in the STUN message. Instead, we can use the remote's fingerprint learned through the DTLS handshake in the following Noise handshake.
> 
> Okay, but what should I use for STUN username and password?
> 
> * If _A_ constructs SDP it pretends to have received from _B_, then it needs _B_'s ICE username and password. Currently they are both set to _B_'s fingerprint. _(We need them to establish ICE connection)_
> 
> * _B_ can use username from STUN binding request (it can be random, BUT we don't know the password, so we'll have to agree that username==password).

From what I can tell, requiring username==password is fine. Still, could you detail why this has to be the case? This is an implementation and not a protocol requirement, correct? In other words, given that we don't actually use the username+password security mechanism, it doesn't add any value, correct? Do I understand correctly that we could as well use a static value for both across all nodes?

> So, while removing the necessity to put fingerprint into STUN binding request simplifies stuff (not a lot, I'd say it complicates it actually because one side will be putting fingerprints, another - random values)

For the STUN message from B->A (i.e. Server->Client) do I understand correctly that static username+password would suffice as well?

> , it does not really solve the problem described in [#412 (comment)](https://github.com/libp2p/specs/pull/412#issuecomment-1170906786).

Correct. Neither putting the fingerprint in the username+password, nor using a static username+password adds any DOS defense. Please let me know in case I am missing something. I don't currently see a way around some rate-limiting mechanism to prevent the DOS attack vector.

_Thanks @melekes for all the help here :pray:_

Reply:
> In other words, given that we don't actually use the username+password security mechanism, it doesn't add any value, correct? Do I understand correctly that we could as well use a static value for both across all nodes?

Current implementation requires unique usernames for each connection. I.e. there's a connections map `var map[username]Connection`. Of course we can set it to "{IP}/{PORT}". But that's an implementation detail. As for security, I will need to read more about ICE username/password to make a judgment. But maybe someone else knows more.

Reply:
@melekes "{IP}/{PORT}" could be problematic in the case of peer-reflexive candidates. On the browser side, you would need the server-reflexive candidate first to set the username as "{IP}/{PORT}" and this will be gathered only after the SDP (local description) is set. Reapplying would result in ice renegotiation and could lead to new candidates.

@mxinden Yes, it is possible to use DTLS in insecure mode.

Reply:
> The DTLS handshake doesn't give us the security properties we want (libp2p peer ID verification) anyway, that's why we're running a Noise handshake later.

If DTLS handshake doesn't give us anything, why not "drop" (use static) server certificate either as suggested in https://github.com/libp2p/specs/pull/412#discussion_r877141696. That way we won't need `certhash` at all!

Reply:
> If DTLS handshake doesn't give us anything

This is not accurate. It does give us something (a lot, actually): a secure connection. The only thing it doesn't tell us if there's a MITM in the connection. By running a Noise handshake on **one stream** we achieve:
1. we learn the peer IDs
2. we tie the DTLS connection (by checking the certificate fingerprints) to a channels that's provably MITM-free

Thereafter, we can use native WebRTC streams without applying any additional encryption.

The proposal to use single, public "private key" (which is what "static server certificate" actually means) gives us neither of these:
* In this case, the DTLS handshake is actually completely useless
* We can't use WebRTC streams, instead we'd have to use Noise on one stream and then apply a stream muxer.
* All the other streams can't be used at all.

Reply:
> (b) exposes the fingerprint of the remote learned in the DTLS handshake to the user (here go-libp2p)?

Looks like I was slightly wrong here. It's not possible to get remote certificate through peer connection (although it's available in DtlsTransport struct). Opened https://github.com/webrtc-rs/webrtc/pull/221. It might also be good to open a similar PR in pion and receive a feedback (since it's more well maintained). @ckousik is that something you'd like to do? Otherwise, I might do that.

Reply:
Opened https://github.com/pion/webrtc/pull/2284